src/header/common/origin.rs

```diff
@@ -31,41 +31,70 @@ use header::parsing::from_one_raw_str;
 /// );
 /// ```
 
-#[derive(Clone, Debug)]
-pub struct Origin {
-    /// The scheme, such as http or https
-    scheme: Cow<'static,str>,
-    /// The host, such as Host{hostname: "hyper.rs".to_owned(), port: None}
-    host: Host,
+#[derive(PartialEq, Clone, Debug)]
+pub struct Origin(OriginOrNull);
+
+#[derive(PartialEq, Clone, Debug)]
+enum OriginOrNull {
+    Origin {
+        /// The scheme, such as http or https
+        scheme: Cow<'static,str>,
+        /// The host, such as Host{hostname: "hyper.rs".to_owned(), port: None}
+        host: Host,
+    },
+    Null,
 }
 
 impl Origin {
     /// Creates a new `Origin` header.
     pub fn new<S: Into<Cow<'static,str>>, H: Into<Cow<'static,str>>>(scheme: S, hostname: H, port: Option<u16>) -> Origin{
-        Origin {
+        Origin(OriginOrNull::Origin {
             scheme: scheme.into(),
             host: Host::new(hostname, port),
+        })
+    }
+
+    /// Creates a `Null` `Origin` header.
+    pub fn null() -> Origin {
+        Origin(OriginOrNull::Null)
+    }
+
+    /// Checks if `Origin` is `Null`.
+    pub fn is_null(&self) -> bool {
+        match self {
+            &Origin(OriginOrNull::Null) => true,
+            _ => false,
         }
     }
 
     /// The scheme, such as http or https
     /// ```
     /// use hyper::header::Origin;
     /// let origin = Origin::new("https", "foo.com", Some(443));
-    /// assert_eq!(origin.scheme(), "https");
+    /// assert_eq!(origin.scheme(), Some("https"));
     /// ```
-    pub fn scheme(&self) -> &str {
-        &(self.scheme)
+    #[allow(unused_variables)]
+    // Variable `host` is considered unused, although there does not seem to be a way to omit it here.
```

Comment:
Another trick is to add `_` to the names of otherwise-unused variables, which suppresses the warning.